### PR TITLE
Make DOM own the style and layout data, in an UnsafeCell

### PR DESCRIPTION
**Pull request by nox**

### CODE REVIEW

File: components/style/invalidation/element/state_and_attributes.rs [hunk 1]
```diff
@@ -90,7 +90,7 @@ pub fn invalidated_descendants<E>(element: E, child: E)
 where
     E: TElement,
 {
-    if child.get_data().is_none() {
+    if child.has_data() {
```
Comment (emilio) [inline components/style/invalidation/element/state_and_attributes.rs:93]: Got the boolean logic wrong on this one.

File: components/layout_2020/traversal.rs [hunk 2]
```diff
@@ -63,7 +63,7 @@ where
     }
 
     fn text_node_needs_traversal(node: E::ConcreteNode, parent_data: &ElementData) -> bool {
-        node.get_raw_data().is_none() || !parent_data.damage.is_empty()
+        (node.get_raw_data().is_none()) || !parent_data.damage.is_empty()
```
Comment (emilio) [inline components/layout_2020/traversal.rs:66]: extra parens for no good reason?

File: components/layout/traversal.rs [hunk 3]
```diff
@@ -73,7 +73,7 @@ where
         // flow construction:
         // (1) They child doesn't yet have layout data (preorder traversal initializes it).
         // (2) The parent element has restyle damage (so the text flow also needs fixup).
-        node.get_raw_data().is_none() || !parent_data.damage.is_empty()
+        (node.get_raw_data().is_none()) || !parent_data.damage.is_empty()
```
Comment (emilio) [inline components/layout/traversal.rs:76]: Extra parens for no good reason?

File: components/script_layout_interface/lib.rs [hunk 4]
```diff
@@ -49,12 +50,32 @@ impl StyleData {
...
 pub struct OpaqueStyleAndLayoutData {
     // NB: We really store a `StyleAndLayoutData` here, so be careful!
-    #[ignore_malloc_size_of = "TODO(#6910) Box value that should be counted but \
-                               the type lives in layout"]
-    pub ptr: NonNull<StyleData>,
+    #[ignore_malloc_size_of = "Trait objects are hard"]
+    ptr: Box<dyn Any + Send + Sync>,
+}
+
+impl OpaqueStyleAndLayoutData {
```
Comment (emilio) [inline components/script_layout_interface/lib.rs:60]: Clever / horrified :)

**Review by emilio — APPROVED** [review]
r=me with the bits below fixed.

**Comment by jdm** [comment]
```
failures:
    size_of::size_characterdata
    size_of::size_div
    size_of::size_element
    size_of::size_htmlelement
    size_of::size_node
    size_of::size_span
    size_of::size_text
```

**Comment by nox** [comment]
Oh right, of course.

**Comment by nox** [comment]
Oops…

**Pull request merged**